clients/filesystem-hadoop/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java

```diff
@@ -0,0 +1,406 @@
+                      }
+                    })
+            .build();
+  }
+
+  @VisibleForTesting
+  Cache<NameIdentifier, FilesetMeta> getFilesetCache() {
+    return filesetCache;
+  }
+
+  @Override
+  public URI getUri() {
+    return this.uri;
+  }
+
+  @Override
+  public FSDataInputStream open(Path f, int bufferSize) throws IOException {
+    NameIdentifier identifier = extractIdentifier(f.toUri());
+    FilesetMeta meta = getCachedFileset(identifier);
+    Path actualPath = resolvePathByIdentifier(identifier, meta, f);
+    return meta.getFileSystem().open(actualPath, bufferSize);
+  }
+
+  @Override
+  public FSDataOutputStream create(
+      Path f,
+      FsPermission permission,
+      boolean overwrite,
+      int bufferSize,
+      short replication,
+      long blockSize,
+      Progressable progress)
+      throws IOException {
+    NameIdentifier identifier = extractIdentifier(f.toUri());
+    FilesetMeta meta = getCachedFileset(identifier);
+    Path actualPath = resolvePathByIdentifier(identifier, meta, f);
+    return meta.getFileSystem()
+        .create(actualPath, permission, overwrite, bufferSize, replication, blockSize, progress);
+  }
+
+  @Override
+  public FSDataOutputStream append(Path f, int bufferSize, Progressable progress)
+      throws IOException {
+    NameIdentifier identifier = extractIdentifier(f.toUri());
+    FilesetMeta meta = getCachedFileset(identifier);
+    Path actualPath = resolvePathByIdentifier(identifier, meta, f);
+    return meta.getFileSystem().append(actualPath, bufferSize, progress);
+  }
+
+  @Override
+  @SuppressWarnings("deprecation")
+  public boolean rename(Path src, Path dst) throws IOException {
+    // There are two cases that cannot be renamed:
+    // 1. Fileset identifier is not allowed to be renamed, only its subdirectories can be renamed
+    // which not in the storage location of the fileset;
+    // 2. Fileset only mounts a single file, the storage location of the fileset cannot be renamed;
+    // Otherwise the metadata in the Gravitino server may be inconsistent.
+    NameIdentifier srcIdentifier = extractIdentifier(src.toUri());
+    NameIdentifier dstIdentifier = extractIdentifier(dst.toUri());
+    Preconditions.checkArgument(
+        srcIdentifier.equals(dstIdentifier),
+        "Destination path fileset identifier: %s should be same with src path fileset identifier: %s.",
+        srcIdentifier,
+        dstIdentifier);
+
+    FilesetMeta meta = getCachedFileset(srcIdentifier);
+    if (meta.getFileSystem().isFile(new Path(meta.getFileset().storageLocation()))) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "Cannot rename the fileset: %s which only mounts a single file.", srcIdentifier));
+    }
+
+    Path srcActualPath = resolvePathByIdentifier(srcIdentifier, meta, src);
+    Path dstActualPath = resolvePathByIdentifier(dstIdentifier, meta, dst);
+    return meta.getFileSystem().rename(srcActualPath, dstActualPath);
```

Comment:
There should be no need to check here, because the `rename` operation only allows `src` and `dst` to correspond to the same `fileset`, which ensures that the `StorageLocation` obtained from the server is the same